common/src/main/java/tachyon/security/PlainSaslServer.java

```diff
@@ -144,4 +148,40 @@ private void throwIfNotComplete() {
       throw new IllegalStateException("PLAIN authentication not completed");
     }
   }
+
+  public static final class PlainServerCallbackHandler implements CallbackHandler {
+    private final AuthenticationMethod mAuthMethod;
+
+    public PlainServerCallbackHandler(String authMethodStr) {
+      mAuthMethod = AuthenticationMethod.getValidAuthenticationMethod(authMethodStr);
+    }
+
+    @Override
+    public void handle(Callback[] callbacks) throws IOException, UnsupportedCallbackException {
+      String username = null;
+      String password = null;
+      AuthorizeCallback ac = null;
+
+      for (Callback callback : callbacks) {
+        if (callback instanceof NameCallback) {
+          NameCallback nc = (NameCallback) callback;
+          username = nc.getName();
+        } else if (callback instanceof PasswordCallback) {
+          PasswordCallback pc = (PasswordCallback) callback;
+          password = new String(pc.getPassword());
+        } else if (callback instanceof AuthorizeCallback) {
+          ac = (AuthorizeCallback) callback;
+        } else {
+          throw new UnsupportedCallbackException(callback);
+        }
+      }
+
+      AuthenticationProvider provider =
+          AuthenticationProviderFactory.getAuthenticationProvider(mAuthMethod);
+      provider.authenticate(username, password);
```

Comment:
I think it would be better to catch `AuthenticationException` in `evaluateResponse` and add more accurate error message, e.g. whatever implemented in `#authenticate`, because eventually we want the server to pass this failure information to the Tachyon client, and the client might not see the stack trace.